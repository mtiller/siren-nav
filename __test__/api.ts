import MockAdapter from "axios-mock-adapter";
import axios from "axios";
import { Siren } from "siren-types";

function s<T>(obj: Siren<T>): Siren<T> {
  return obj;
}

export function usingMockAPI(tests: (mock: MockAdapter) => Promise<void>) {
  return async (): Promise<void> => {
    const mock = new MockAdapter(axios);

    //// Resources that deal with redirection and location headers
    mock.onGet("http://localhost/308").reply(308, "Wrong place, move along", {
      Location: "http://localhost/bar"
    });
    mock.onGet("http://localhost/resource").reply(
      200,
      {
        message: "You really should follow my location header"
      },
      {
        Location: "http://localhost/bar"
      }
    );
    mock.onGet("http://localhost/bar").reply(
      200,
      s({
        properties: {
          message: "I am bar"
        }
      })
    );
    mock.onGet("http://localhost/array").reply(200, new ArrayBuffer(10));
    mock.onGet("http://localhost/string").reply(200, "a string");
    mock.onGet("http://localhost/string").reply(200, 5);

    //// Resources that involve templates
    mock.onGet("http://localhost/search-template").reply(
      200,
      s({
        properties: undefined,
        links: [{ rel: ["search"], href: "/model/{model}" }]
      })
    );

    //// Resources that provide collections
    mock.onGet("http://localhost/collection").reply(
      200,
      s({
        properties: undefined,
        links: [
          { rel: ["item"], href: "/model/1" },
          { rel: ["item"], href: "/model/2" }
        ]
      })
    );
    mock.onGet("http://localhost/model/1").reply(
      200,
      s({
        properties: { id: 1 },
        links: [{ rel: ["collection"], href: "/collection" }]
      })
    );
    mock.onGet("http://localhost/model/2").reply(
      200,
      s({
        properties: { id: 2 },
        links: [{ rel: ["collection"], href: "/collection" }]
      })
    );

    //// A self referential resource
    mock.onGet("http://localhost/self-ref").reply(
      200,
      s({
        properties: undefined,
        links: [{ rel: ["self"], href: "http://localhost/self-ref" }]
      })
    );

    mock.onGet("http://localhost/api").reply(
      200,
      s({
        properties: {
          queryString: false
        },
        actions: [
          {
            name: "query",
            href: "/api/query"
          },
          {
            name: "create",
            href: "/api/create",
            method: "POST",
            type: "application/json"
          }
        ]
      })
    );

    mock.onGet("http://localhost/api/query").reply(
      200,
      s({
        properties: {
          case: 1,
          queryString: false,
          flattened: false
        },
        links: []
      })
    );

    mock.onGet("http://localhost/api/query?term=home&x=5").reply(
      200,
      s({
        properties: {
          case: 2,
          queryString: true,
          flattened: false
        },
        links: []
      })
    );

    mock.onGet("http://localhost/api/query?parent.child=5").reply(
      200,
      s({
        properties: {
          case: 3,
          queryString: true,
          flattened: true
        },
        links: []
      })
    );

    mock.onGet("http://localhost/api/query?parent=%7B%22child%22%3A5%7D").reply(
      200,
      s({
        properties: {
          case: 4,
          queryString: true,
          flattened: false
        },
        links: []
      })
    );

    mock
      .onGet(
        "http://localhost/api/query?properties.x=5&links.0.rel.0=item&links.0.href=%2Ffoo"
      )
      .reply(
        200,
        s({
          properties: {
            case: 5,
            queryString: true,
            flattened: true
          },
          links: []
        })
      );

    mock.onPost("http://localhost/api/create").reply(
      200,
      s({
        properties: undefined
      })
    );
    const msg1: Siren<{ message: string }> = {
      properties: { message: "My message" }
    };
    const msg2: Siren<{ message: string }> = {
      properties: { message: "Another message" }
    };
    mock
      .onGet("http://localhost/sse")
      .reply(
        200,
        `data: ${JSON.stringify(msg1)}\n\ndata: ${JSON.stringify(msg2)}\n\n`,
        {
          "Content-Type": "text/event-stream"
        }
      );

    try {
      await tests(mock);
    } catch (e) {
      throw e;
    } finally {
      mock.reset();
    }
  };
}
