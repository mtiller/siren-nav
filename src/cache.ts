import { Entity } from "siren-types";

export class Cache {
    private cache: { [key: string]: Entity<any> };
    constructor() {
        this.cache = {};
    }
    add(uri: string, resp: Entity<any>) {
        this.cache[uri] = resp;
    }
    has(uri: string): boolean {
        return this.cache.hasOwnProperty(uri);
    }
    get(uri: string): Entity<any> {
        if (this.has(uri)) return this.cache[uri];
        throw new Error("Requested uncached resource: " + uri);
    }
    getOr(uri: string): Entity<any> | null {
        if (this.has(uri)) return this.cache[uri];
        return null;
    }
}
