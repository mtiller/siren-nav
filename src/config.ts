import { AxiosRequestConfig } from "axios";

import * as debug from "debug";
const debugConfig = debug("siren-nav:config");

export type Config = AxiosRequestConfig;
export type Configurator = (cur: Config) => Config;

export const headerConfig = (key: string, value: string): Configurator => {
    return (config: Config) => {
        let newconfig = { ...config };
        if (!newconfig.headers) newconfig.headers = {};

        // NB - We must create a new header object!
        newconfig.headers = { ...config.headers };
        newconfig.headers[key] = value;

        debugConfig("  Set header '%s' to '%s'", key, newconfig.headers[key]);
        return newconfig;
    };
};

export const contentTypeConfig = (value: string): Configurator => {
    return headerConfig("Content-Type", value);
};

export const acceptConfig = (value: string): Configurator => {
    return headerConfig("Accept", value);
};
