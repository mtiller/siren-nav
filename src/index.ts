export * from "./navigation";
export * from "./requests";
export * from "./response";
export * from "./state";
export * from "./steps/apply";
export * from "./utils";
