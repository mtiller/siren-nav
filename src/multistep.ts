import { NavState } from "./state";
import { getSiren } from "./utils";
import { findPossible, Step } from "./steps";
import { Cache } from "./cache";

export type MultiStep = (cur: NavState, cache: Cache) => Promise<NavState[]>;

export function followEach(rel: string, parameters: {} | undefined): MultiStep {
    return (state: NavState, cache: Cache): Promise<NavState[]> => {
        return getSiren(state).then(siren => {
            const possible = findPossible(rel, siren, state, cache, parameters);
            return possible;
        });
    };
}

export function toMulti(step: Step): MultiStep {
    return async (cur: NavState, cache: Cache) => {
        const ns: NavState = await step(cur, cache);
        return [ns];
    };
}
