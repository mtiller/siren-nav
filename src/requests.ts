import { NavState } from "./state";
import { Siren } from "siren-types";
import axios from "axios";
import { AxiosResponse } from "axios";

import { flatten } from "flat";

import URI from "urijs";

import debug from "debug";
import { normalizeUrl, urlEncoded } from "./utils";
import { headerConfig, Config } from "./config";
const log = debug("siren-nav:requests");

export interface ResponseData {
  data: any;
  headers: {};
  status: number;
}

export interface ActionOptions {
  flatten?: boolean;
}

export type Request = (cur: NavState) => Promise<ResponseData>;

export function performAction<T extends {}>(
  name: string,
  body: T,
  config: Config,
  parameters?: {}
): Request {
  return async (state: NavState): Promise<AxiosResponse> => {
    log("Performing action %s on %s", name, state.cur);
    log("  Fetching with current state of: %j", state);
    const resp = await getRequest(state);
    let siren: Siren = resp.data as Siren;
    // We may need to tweak the config, so let's start with the config of
    // the NavState
    let request = state.currentConfig;

    log("  Latest value of %j: %j", siren, state.cur);

    if (!siren.actions) {
      log("  ERROR: no actions defined in %s", state.cur);
      throw new Error("No actions defined for " + state.cur);
    }

    const action = siren.actions.find(action => action.name == name);
    if (!action) {
      throw new Error(
        `No action named ${name} found among: ${JSON.stringify(
          siren.actions.map(action => action.name)
        )}`
      );
    }
    log("  Found action with name %s", name);

    // Make sure there is an href here...
    if (!action.href) {
      log("    ERROR: no href");
      throw new Error("No href for action " + name);
    }

    // If the action specifies a content-type for the payload, include that
    // in the request.
    if (action.type) {
      log("    Content-Type set to %s", action.type);
      request = headerConfig("Content-Type", action.type)(request);
    }

    const method = (action.method || "get").toLowerCase();
    // Specify the method associated with the action (or fallback to "get")
    // Note that axios expects "get" or "post" (i.e., no GET, no POST).
    request = { ...request, method: method };

    let url = normalizeUrl(action.href, state.cur, parameters);

    if (urlEncoded(action)) {
      if (typeof body === "string") {
        request = { ...request, url: url + body };
        log("  Encoding data with provided query string: %s", body);
      } else {
        if (config.flatten) {
          const fqs = flatten(body) as {};
          log("  Encoding data as flattened query string using: %j", fqs);
          const qs = new URI("").search(fqs).toString();
          log("  Query string being used: %s", qs);
          request = { ...request, url: url + qs };
        } else {
          log("  Encoding data with unflattened query string using: %j", body);
          const params = Object.keys(body).reduce((p, val) => {
            p[val] = JSON.stringify(body[val]);
            return p;
          }, {} as { [key: string]: string });
          const qs = new URI("").search(params).toString();
          log("  Query string being used: %s", qs);
          request = { ...request, url: url + qs };
        }
      }
    } else {
      // Add URL and data to request
      log("  Encoding data as JSON in body: %j", body);
      request = { ...request, url: url, data: body };
    }

    log("    Making a request to %s with config %j", url, request);
    return Promise.resolve(axios.request(request));
  };
}

export const getRequest: Request = async (
  state: NavState
): Promise<ResponseData> => {
  log("Requesting %s, current state is: %j", state.cur, state);
  log(
    "  Fetching latest resource at %s with config %j",
    state.cur,
    state.currentConfig
  );
  return axios.get(state.cur, state.currentConfig).then(v => {
    log("  ResponseData for %s: %j", state.cur, v);
    return v;
  });
};
