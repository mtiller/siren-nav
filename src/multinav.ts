import { NavState } from "./state";
import { MultiStep, toMulti, followEach } from "./multistep";
import { Step, reduceEach, follow } from "./steps";
import { MultiResponse } from "./response";
import { getRequest } from "./requests";
import { Cache } from "./cache";

/**
 * The MultiNav class is not one a user would generally instantiate themselves.
 * Instead, it is created in response to certain steps in a chain that cause the
 * set of resources to "fan out".  The main example of this is when you follow a
 * relation and there are multiple instances of that relation.
 *
 * @export
 * @class MultiNav
 */
export class MultiNav {
    constructor(
        private start: Promise<NavState[]>,
        private steps: MultiStep[],
        private omni: Step[],
        private cache: Cache,
    ) {}
    get(): MultiResponse {
        let state = reduceEach(this.start, [...this.steps, ...this.omni.map(toMulti)], this.cache);
        let resp = state.then(s => Promise.all(s.map(x => getRequest(x))));
        return MultiResponse.create(resp);
    }
    do(step: Step): MultiNav {
        return new MultiNav(this.start, [...this.steps, toMulti(step)], this.omni, this.cache);
    }
    doMulti(steps: MultiStep): MultiNav {
        return new MultiNav(this.start, [...this.steps, steps], this.omni, this.cache);
    }

    follow(rel: string, parameters?: {}, which?: (states: NavState[]) => NavState): MultiNav {
        return this.do(follow(rel, parameters, which));
    }

    followEach(rel: string, parameters?: {}): MultiNav {
        return this.doMulti(followEach(rel, parameters));
    }
}
