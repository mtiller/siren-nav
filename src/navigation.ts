import { NavState, Config } from "./state";
import { Cache } from "./cache";
import { follow, Step, reduce, accept, auth, followLocation, header } from "./steps";
import { Entity } from "siren-types";
import { NavResponse } from "./response";
import { performAction, getRequest } from "./requests";
import { sirenContentType } from "siren-types";
import { Observable } from "rxjs";
import { MultiNav } from "./multi/multinav";

import * as URI from "urijs";
import { followEach, toMulti } from "./multi/multistep";

/**
 * The SirenNav class provides a collection of methods that allow for
 * a high-level declarative based approach to API navigation.
 *
 * @export
 * @class SirenNav
 */
export class SirenNav {
    /**
     * Create a new SirenNav instance for an API with an entry point
     * specified by the argument.
     *
     * @static
     * @param {string} url Initial URL
     * @param {string} base Base URL for API (all URIs will be anchored to this)
     * @returns
     *
     * @memberOf SirenNav
     */
    static create(url: string, cache?: Cache, config?: Config) {
        if (!cache) cache = new Cache();
        config = config || {};
        const uri = URI(url);
        if (uri.is("relative")) {
            throw new Error("SirenNav must be created with an absolute URL");
        }
        return new SirenNav(
            Promise.resolve(
                new NavState(
                    url,
                    undefined,
                    {
                        headers: {
                            Accept: sirenContentType, // Assume siren unless the user overrides it
                        },
                        // Commented out because it causes problems with HTTPS
                        // APIs that don't require authentication (for reasons
                        // I'm not 100% sure about).
                        // withCredentials: true,
                        ...config,
                    },
                    cache.getOr(url),
                ),
            ),
            [],
            [],
            cache,
        );
    }

    /**
     * Creates an instance of SirenNav.  This constructor is private because
     * it initializes lots of state that should be encapsulated (i.e., we
     * don't generally want people mucking about with this stuff).  Instead,
     * we use static functions to create instances.
     *
     * @param {Promise<NavState>} start
     * @param {Step[]} steps
     *
     * @memberOf SirenNav
     */
    private constructor(
        private start: Promise<NavState>,
        private steps: Step[],
        private omni: Step[],
        private cache: Cache,
    ) {}

    /**
     * Follow **one** instance of a given relation.  The optional second
     * argument is used to indicate that the first match should be used in cases
     * where multiple matches are available.  If the second argument is
     * undefined or false, then anything but a single (exact) match will result
     * in an error.
     *
     * If you wish to follow all links of a given relation, use followEach.
     *
     * @param {string} rel
     * @param {boolean} [first]
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    follow(rel: string, parameters?: {}, which?: (states: NavState[]) => NavState): SirenNav {
        return this.do(follow(rel, parameters, which));
    }

    /**
     * Follow the location header in the response when fetching the current resource.
     *
     * @returns {SirenNav}
     * @memberof SirenNav
     */
    followLocation(): SirenNav {
        return this.do(followLocation);
    }

    followEach(rel: string, parameters?: {}): MultiNav {
        const multi = this.asMultiNav();
        return multi.doMulti(followEach(rel, parameters));
    }

    /**
     * Perform an action that expects a "hypermedia" payload as the input
     * to the function.
     *
     * @template P
     * @param {string} name
     * @param {Entity<P>} body
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    performHyperAction<P extends {}>(name: string, body: Entity<P>): NavResponse {
        return this.performAction(name, body);
    }

    /**
     * Perform an action using an arbitrary type of input data.
     *
     * @template P
     * @param {string} name
     * @param {P} body
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    performAction<P>(name: string, body: P): NavResponse {
        let state = reduce(this.start, [...this.steps, ...this.omni], this.cache);
        let resp = state.then(s => performAction(name, body)(s));
        return NavResponse.create(resp);
    }

    /**
     * Perform a navigation step.  This is mainly an internal method, but it
     * is exposed as a public method as to allow the set of possible
     * navigation steps to be extensible.  Any transformation satisfying
     * the Step type can be incorporated into the navigation process.
     *
     * @param {Step} step
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    do(step: Step): SirenNav {
        return new SirenNav(this.start, [...this.steps, step], [...this.omni], this.cache);
    }

    /**
     * Perform a step that should *always* be performed.  These omni steps are done
     * after the navigation steps.  This is also an internal method, but it is made
     * public to allow extensions.
     */
    doOmni(step: Step): SirenNav {
        return new SirenNav(this.start, [...this.steps], [...this.omni, step], this.cache);
    }

    /**
     * This method can be used to generate a new SirenNav instance that
     * has already prenavigated a particular set of steps.  When a
     * "normal" SirenNav instance is evaluated, all the steps in the
     * navigation are repeated.  If the same navigation is repeated
     * over and over, many HTTP requests can potentially be made.
     * If those requests are sure to always return the same result,
     * then these requests will cause delays and unnecessarily
     * network requests.  The squash method performs all the
     * specified steps of the navigation BUT uses the resulting
     * state to create a new SirenNav instance that always starts
     * at the resulting state.  It therefore avoids repeating
     * steps on each execution.  However, squashing a SirenNav
     * is only appropriate if the preceding steps are sure to
     * always return the same results.
     *
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    squash(): SirenNav {
        // NB - Note that we do NOT squash the omni steps.  Those remain.
        return new SirenNav(reduce(this.start, this.steps, this.cache), [], [...this.omni], this.cache);
    }

    /**
     * This tells the navigator to jump to a given URL.  This is
     * different from creating a new navigator because it
     * retains any information about the root URI of the API
     * and any configuration changes that have accumulated.
     *
     * @param {string} url
     * @returns {SirenNav}
     *
     * @memberOf SirenNav
     */
    goto(url: string, parameters?: {}): SirenNav {
        let newstate = new Promise<NavState>(async (resolve, reject) => {
            let state = await this.start;
            resolve(new NavState(state.rebase(url), parameters, state.config, this.cache.getOr(url)));
        });
        return new SirenNav(newstate, [], [...this.omni], this.cache);
    }

    /**
     * Add an entry to the Accept header.  This can be called multiple times and each
     * call will add an additional option.  As with the Accept header itself, order
     * matters.
     */
    accept(ctype: string): SirenNav {
        return this.do(accept(ctype));
    }

    /**
     * Authorize requests using the given scheme and token
     */
    auth(scheme: string, token: string): SirenNav {
        return this.do(auth(scheme, token));
    }

    /**
     * Add a given header
     *
     * @param {string} key The name of the header
     * @param {string} value The value to give to the header
     * @returns {SirenNav}
     * @memberof SirenNav
     */
    header(key: string, value: string): SirenNav {
        return this.do(header(key, value));
    }

    /**
     * This method returns the URL of the resource that is being pointed
     * to at the end of the chain of navigation steps.
     *
     * @returns
     *
     * @memberOf SirenNav
     */
    getURL(): Promise<string> {
        return reduce(this.start, [...this.steps, ...this.omni], this.cache).then(state => {
            if (state.config.baseURL) {
                return URI(state.cur)
                    .absoluteTo(state.config.baseURL)
                    .toString();
            }
            return state.cur;
        });
    }

    /**
     * The get method returns a NavResponse the represents the value of
     * the current resource.
     *
     * @returns {NavResponse}
     *
     * @memberOf SirenNav
     */
    get(): NavResponse {
        let state = reduce(this.start, [...this.steps, ...this.omni], this.cache);
        let resp = state.then(s => getRequest(s));
        return NavResponse.create(resp);
    }

    // TODO: Test this
    subscribe<T>(rel?: string, poll?: number): Observable<Entity<T>> {
        return new Observable(observer => {
            observer.complete();
            (async () => {
                // Follow rel
                // Get URL
                const url = await this.follow(rel || "events").getURL();
                if (url.startsWith("ws:" || url.startsWith("wss:"))) {
                    throw new Error("Unimplemented");
                    // Handle web socket case
                } else {
                    // Handle event source case
                    const source = new EventSource(url);
                    source.onmessage = msg => observer.next(msg as any);
                    // TODO: Distinguish between close and error
                    source.onerror = e => observer.error(e);
                }
            })();
        });
    }

    /**
     * Transforms the current SirenNav into a MultiNav (one capable of tracking
     * multiple concurrent resources)
     *
     * @protected
     * @returns {MultiNav}
     * @memberof SirenNav
     */
    protected asMultiNav(): MultiNav {
        const starts: Promise<NavState[]> = this.start.then(v => [v]);
        return new MultiNav(starts, this.steps.map(toMulti), this.omni, this.cache);
    }
}
