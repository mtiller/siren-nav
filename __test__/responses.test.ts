import { usingMockAPI } from "./api";
import { SirenNav } from "../src/navigation";

describe("Response Tests", () => {
    it(
        "should follow a template relation",
        usingMockAPI(async () => {
            const nav = SirenNav.create("http://localhost/bar");
            const resp = await nav.get();
            const siren = await resp.asSiren();
            expect(siren).toMatchSnapshot();
            const nsiren = await resp.asNormalizedSiren();
            expect(nsiren).toMatchSnapshot();
            const json = await resp.asJson();
            expect(json).toMatchSnapshot();
            const buffer = await resp.asBuffer();
            expect(buffer).toMatchSnapshot();
            const raw = await resp.asRaw();
            expect(raw).toMatchSnapshot();
        }),
    );
    it(
        "should handle arrays",
        usingMockAPI(async () => {
            const nav = SirenNav.create("http://localhost/array");
            const resp = await nav.get();
            const buffer = await resp.asBuffer();
            expect(typeof buffer).toEqual("object");
            expect(buffer).toMatchSnapshot();
        }),
    );
    it(
        "should handle strings",
        usingMockAPI(async () => {
            const nav = SirenNav.create("http://localhost/string");
            const resp = nav.get();
            expect(resp.asBuffer()).rejects.toMatchSnapshot();
        }),
    );
    it(
        "should handle number",
        usingMockAPI(async () => {
            const nav = SirenNav.create("http://localhost/number");
            const resp = nav.get();
            expect(resp.asBuffer()).rejects.toMatchSnapshot();
        }),
    );
});
