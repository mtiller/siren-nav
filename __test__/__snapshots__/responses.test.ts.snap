// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Response Tests should follow a template relation 1`] = `
Object {
  "properties": Object {
    "message": "I am bar",
  },
}
`;

exports[`Response Tests should follow a template relation 2`] = `
Object {
  "actions": Array [],
  "class": Array [],
  "entities": Array [],
  "links": Array [],
  "title": null,
}
`;

exports[`Response Tests should follow a template relation 3`] = `
Object {
  "properties": Object {
    "message": "I am bar",
  },
}
`;

exports[`Response Tests should follow a template relation 4`] = `"{\\"properties\\":{\\"message\\":\\"I am bar\\"}}"`;

exports[`Response Tests should follow a template relation 5`] = `
Object {
  "config": Object {
    "data": undefined,
    "flatten": true,
    "headers": Object {
      "Accept": "application/vnd.siren+json",
    },
    "maxContentLength": -1,
    "method": "get",
    "timeout": 0,
    "transformRequest": Object {
      "0": [Function],
    },
    "transformResponse": Object {
      "0": [Function],
    },
    "url": "http://localhost/bar",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": Object {
    "properties": Object {
      "message": "I am bar",
    },
  },
  "headers": undefined,
  "status": 200,
}
`;

exports[`Response Tests should handle arrays 1`] = `ArrayBuffer []`;

exports[`Response Tests should handle number 1`] = `[Error: Request failed with status code 404]`;

exports[`Response Tests should handle number 2`] = `[Error: Request failed with status code 404]`;
